### PR TITLE
[remotetapprocessor] use 'time/rate' to limit traffic

### LINKED ISSUE #32385: RemoteTap Processor doesn't support a limit
### Component(s)

processor/remotetap

## Description
The remote tap processor specifies a [limit](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/remotetapprocessor/config.go#L21), however, the limit is never used.

## Steps to Reproduce
Specify a limit on a busy pipeline and be bombarded with messages.

## Expected Result
The limit is applied to not overwhelm a consumer

## Actual Result
The consumer is overwhelmed.

### Collector version

0.98.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 127.0.0.1:4317
      http:
        endpoint: 127.0.0.1:4318

processors:
  remotetap:
    endpoint: 127.0.0.1:12001
    limit: 1
  batch:

exporters:
  debug:
    verbosity: detailed

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [remotetap, batch]
      exporters: [debug]

    metrics:
      receivers: [otlp]
      processors: [remotetap, batch]
      exporters: [debug]

    logs:
      receivers: [otlp]
      processors: [remotetap, batch]
      exporters: [debug]
```

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ahern-pro**
**Description:** <Describe what has changed.>

bug: The remotetapprocessor `limit` configure doesn't work.
how to fix: use `time/rate` to limit traffic. 

**Link to tracking Issue:** <Issue number if applicable>
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32385

### CODE REVIEW

File: processor/remotetapprocessor/README.md [hunk 1]
```diff
@@ -24,8 +24,8 @@ any open WebSockets is rate limited by an adjustable amount.
 
 The WebSocket processor has two configurable fields: `port` and `limit`:
 
-- `port`: The port on which the WebSocket processor listens. Optional. Defaults
-  to `12001`.
+- `endpoint`: The endpoint on which the WebSocket processor listens. Optional. Defaults
+  to `0.0.0.0:12001`.
```
Comment (atoulme) [inline processor/remotetapprocessor/README.md:28]: was this change related to the bug fix?

File: processor/remotetapprocessor/processor.go [hunk 2]
```diff
@@ -12,13 +12,15 @@ import (
 	"sync"
 	"time"
 
+	"go.uber.org/zap"
+	"golang.org/x/net/websocket"
```
Comment (atoulme) [inline processor/remotetapprocessor/processor.go:16]: Probably will fail goimports?

**Comment by atoulme** [comment]
I appreciate your contribution, it looks good with a few comments. Please review.

File: processor/remotetapprocessor/processor.go (hunk 2 above)
Comment (crobert-1) [reply processor/remotetapprocessor/processor.go:16]: Lint failed as @atoulme pointed out:

```
remotetapprocessor/processor.go:15: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
level=info msg="File cache stats: 12 entries of total size 23.4KiB"
level=info msg="Memory: 21 samples, avg is 153.0MB, max is 289.6MB"
level=info msg="Execution took 1.91397649s"
	"go.uber.org/zap"
	"golang.org/x/net/websocket"
	"golang.org/x/time/rate"

remotetapprocessor/processor.go:23: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"go.opentelemetry.io/collector/processor"
make[2]: *** [../../Makefile.Common:199: lint] Error 1
```

File: processor/remotetapprocessor/README.md (hunk 1 above)
Comment (li-zeyuan) [reply processor/remotetapprocessor/README.md:28]: Document error. Has nothing to do with bug fixes.

File: processor/remotetapprocessor/processor.go (hunk 2 above)
Comment (li-zeyuan) [reply processor/remotetapprocessor/processor.go]: Thank you for your review. 
goimports format issue, fixed.

File: processor/remotetapprocessor/README.md (hunk 1 above)
Comment (atoulme) [reply processor/remotetapprocessor/README.md:28]: It can be a separate PR if needed.

**Review by atoulme — APPROVED** [review]

File: processor/remotetapprocessor/README.md (hunk 1 above)
Comment (li-zeyuan) [reply processor/remotetapprocessor/README.md]: Okay, a new PR  for `README.md`

**Comment by li-zeyuan** [comment]
Hi @codeboten  @atoulme @crobert-1, Please review, thanks.

**Comment by crobert-1** [comment]
Hello @li-zeyuan, can you update your PR description so that it will [automatically close the issue when it gets merged](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#description-guidelines)?

File: processor/remotetapprocessor/processor.go [hunk 3]
```diff
@@ -89,31 +92,40 @@ func (w *wsprocessor) Shutdown(ctx context.Context) error {
 }
 
 func (w *wsprocessor) ConsumeMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
-	b, err := metricMarshaler.MarshalMetrics(md)
-	if err != nil {
-		w.telemetrySettings.Logger.Debug("Error serializing to JSON", zap.Error(err))
-	} else {
-		w.cs.writeBytes(b)
+	if w.limiter.Allow() {
```
Comment (crobert-1) [inline processor/remotetapprocessor/processor.go:95]: Is there a way to add a test to ensure that the limit is being applied properly?

File: .chloggen/fix-remotetap-limit.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
```
Comment (crobert-1) [inline .chloggen/fix-remotetap-limit.yaml:4]: I'm a bit concerned about user impact of enabling this. Even though it's clearly documented, users may not be expecting the limit when upgrading here. Should we consider this a breaking change, or assume users will clearly understand what to do here when their perceived performance decreases?

**Review by crobert-1 — APPROVED** [review]

**Comment by li-zeyuan** [comment]
> Hello @li-zeyuan, can you update your PR description so that it will [automatically close the issue when it gets merged](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#description-guidelines)?

ok

File: .chloggen/fix-remotetap-limit.yaml (hunk 4 above)
Comment (li-zeyuan) [reply .chloggen/fix-remotetap-limit.yaml:4]: > I'm a bit concerned about user impact of enabling this. Even though it's clearly documented, users may not be expecting the limit when upgrading here. Should we consider this a breaking change, or assume users will clearly understand what to do here when their perceived performance decreases?

Thanks for your review. You were right to worry. I've labeled it breaking. I can't think of a better way to inform users about the performance degradation after the upgrade, other than documentation.

File: processor/remotetapprocessor/processor.go (hunk 3 above)
Comment (li-zeyuan) [reply processor/remotetapprocessor/processor.go:95]: Have already solved.

**Comment by li-zeyuan** [comment]
Hi @crobert-1 , please review again.

File: .chloggen/fix-remotetap-limit.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Make the `limit` configuration work properly.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [32385]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
[the remark this answers is not in the document]
Comment (li-zeyuan) [reply .chloggen/fix-remotetap-limit.yaml]: @crobert-1 
Thank you for your review, the subtext has been added.

**Review by crobert-1 — APPROVED** [review]

**Pull request merged**